Rerun init containers when the pod needs to be restarted

yujuhong opened the pull request:
Whenever pod sandbox needs to be recreated, all containers associated
with it will be killed by kubelet. This change ensures that the init
containers will be rerun in such cases.

The change also refactors the compute logic so that the control flow of
init containers act is more aligned with the regular containers. Unit
tests are added to verify the logic.

This fixes #36485

yujuhong committed:
Re-run init containers if the pod sandbox needs to be recreated

Whenever pod sandbox needs to be recreated, all containers associated
with it will be killed by kubelet. This change ensures that the init
containers will be rerun in such cases.

The change also refactors the compute logic so that the control flow of
init containers act is more aligned with the regular containers. Unit
tests are added to verify the logic.

yujuhong commented:
/cc @yifan-gu 
Still WIP; not a large-scale improvement, but at least this PR should make init containers work more similar to regular containers.

yujuhong commented:
/cc @smarterclayton

yujuhong commented:
/cc @dchen1107

dchen1107 commented:
@yujuhong if @smarterclayton couldn't review it before 1.7 release, I am going review this pr right after 1.7 release. Is that ok with you?

yujuhong commented:
> @yujuhong if @smarterclayton couldn't review it before 1.7 release, I am going review this pr right after 1.7 release. Is that ok with you?

This is for 1.8, so there is no rush.

smarterclayton commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -394,141 +388,112 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+	return pod.Spec.RestartPolicy != v1.RestartPolicyNever
 }
 
 // computePodContainerChanges checks whether the pod spec has changed and returns the changes if true.
```
rename this as well

yujuhong replied:
Done

smarterclayton commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -394,141 +388,112 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+			restart = true
+		} else if liveness, found := m.livenessManager.Get(containerStatus.ID); found && liveness == proberesults.Failure {
+			// If the container failed the liveness probe, we should kill it.
+			reason = "container failed liveness probe."
```
Should be capitalized like the other reasons

yujuhong replied:
Done

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -660,6 +655,36 @@ func (m *kubeGenericRuntimeManager) pruneInitContainersBeforeStart(pod *v1.Pod,
+			}
+			// Purge all init containers that match this container name
+			glog.V(4).Infof("Removing init container %q instance %q %d", status.Name, status.ID.ID, count)
+			if err := m.runtimeService.RemoveContainer(status.ID.ID); err != nil {
```
Call `m.DeleteContainer()` instead, which also removes logs.

yujuhong replied:
Good call. Updated.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -687,20 +639,10 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
-	if !done {
-		// init container still running
-		glog.V(4).Infof("An init container is still running in pod %v", format.Pod(pod))
-		return
```
Why this is removed? kubelet should wait all init containers finished before starting regular containers.

yujuhong replied:
The `computePodActions` will NOT add any containers to start if init containers are still running. This way, we separate the compute logic and the actual operations.
You can check the unit tests and see those are included in the test cases.

feiskyer replied:
yep, thanks for explanation.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:
```diff
@@ -625,53 +659,396 @@ func TestSyncPodWithInitContainers(t *testing.T) {
+		mutatePodFn    func(*v1.Pod)
+		mutateStatusFn func(*kubecontainer.PodStatus)
+		actions        podActions
+	}{
```
@smarterclayton @dchen1107, please read the description of the test cases and make sure they meet your expectation. Thanks!

yujuhong commented:
/retest

feiskyer commented:
LGTM

munnerz commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -638,19 +638,44 @@ func (m *kubeGenericRuntimeManager) pruneInitContainersBeforeStart(pod *v1.Pod,
+		initContainerNames.Insert(container.Name)
+	}
+	for name := range initContainerNames {
+		count := 0
```
From what I can see here, this 'count' variable never increments? It's printed below on L673, and in the previous implementation would be incremented on each iteration of the `podStatus.ContainerStatuses` loop.

I'm not particularly familiar with this bit of code, so I could be wrong. I'm currently attempting to port this patch to the 1.6 `dockertools` implementation.

yujuhong replied:
Yes, the `count` should be incremented. Will do this later, after getting more comments. The variable is only used in the log message, so it does not affect the functionality of the code.

yujuhong commented:
@dchen1107 @Random-Liu @smarterclayton PTAL.

smarterclayton commented:
Tests look good to me.  This is generally LGTM, although I can do a deeper pass if necessary.

yujuhong commented:
> Tests look good to me. This is generally LGTM, although I can do a deeper pass if necessary.

Thanks for the review! That's exactly what I needed from you and @dchen1107 

@feiskyer already took a pass and LGTM'd the PR. I'll keep the PR open for a couple more days because @Random-Liu mentioned that he may have time to look at it.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -394,141 +388,112 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
-	for index, container := range pod.Spec.Containers {
+	for idx, container := range pod.Spec.Containers {
 		containerStatus := podStatus.FindContainerStatusByName(container.Name)
+		// If container does not exist, or is not running, check whehter we
```
s/whehter/whether

yujuhong replied:
Fixed.

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -394,141 +388,112 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *v1.Pod, podStatus *ku
+			reason = "Container failed liveness probe."
+		} else {
+			// Keep the container.
+			keepCount += 1
```
nit: `++`

yujuhong replied:
I think both are perfectly fine unless this documented as convention(?)

Random-Liu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -587,7 +556,9 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, _ v1.PodStatus, podStat
-	m.pruneInitContainersBeforeStart(pod, podStatus, podContainerChanges.InitContainersToKeep)
+	// This is an optmization because container removals are typically handled
+	// by container garbage collector.
+	m.pruneInitContainersBeforeStart(pod, podStatus)
```
We purge and prune, add an options seems cleaner.

yujuhong replied:
Still think it's clear to keep both functions.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -620,7 +620,7 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, ru
 // of outstanding init containers still present. This reduces load on the container garbage collector
 // by only preserving the most recent terminated init container.
-func (m *kubeGenericRuntimeManager) pruneInitContainersBeforeStart(pod *v1.Pod, podStatus *kubecontainer.PodStatus, initContainersToKeep map[kubecontainer.ContainerID]int) {
+func (m *kubeGenericRuntimeManager) pruneInitContainersBeforeStart(pod *v1.Pod, podStatus *kubecontainer.PodStatus) {
```
I disagree. I think it's clear what each function does with the current naming. Adding a boolean would make the function more complicated and harder to test.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -638,19 +638,44 @@ func (m *kubeGenericRuntimeManager) pruneInitContainersBeforeStart(pod *v1.Pod,
-			if _, ok := initContainersToKeep[status.ID]; ok {
+			// prune all other init containers that match this container name
+			glog.V(4).Infof("Removing init container %q instance %q %d", status.Name, status.ID.ID, count)
+			if err := m.DeleteContainer(status.ID); err != nil {
```
You're right. Switching both to `m.removeContainer(status.ID.ID)` to make them consistent. I think we do need to clean up the log here since we don't do it anywhere else.

yujuhong commented:
@Random-Liu comments addressed. PTAL again, thanks

yujuhong commented:
Pushed a new commit to address a bug @Random-Liu found (good catch!). Also added a new unit test for it.

Random-Liu commented:
@yujuhong LGTM without full confidence. I could understand the current behavior, and it looks good. However, It's hard to figure out the original behavior, so I don't have full confidence that there won't be regression or behavior change.

yujuhong commented:
Rebased and squashed the commits. 
Applying the lgtm label based on LGTMs from https://github.com/kubernetes/kubernetes/pull/47599#issuecomment-322915825 and https://github.com/kubernetes/kubernetes/pull/47599#issuecomment-322335233

feiskyer commented:
@yujuhong Adds a release-note for this?
